app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java

```diff
@@ -89,12 +98,45 @@ public void onClick(View view) {
                     return;
                 }
 
-                _progressBar.setVisibility(View.VISIBLE);
+                checkForUniqueUsername(isUsernameTaken -> {
+                    if (isUsernameTaken) {
+                        Log.d(TAG, "This username already exists");
+
+                        usernameInput.setError(THIS_USERNAME_IS_ALREADY_TAKEN_MESSAGE);
+                        usernameInput.requestFocus();
+
+                        _progressBar.setVisibility(View.INVISIBLE);
+                    } else {
+                        Log.d(TAG, "Username is unique!");
+
+                        createNewUserWithEmailAndPassword(validator.getValidEmailForSignUp(),
+                                                          validator.getValidPasswordForSignUp(),
+                                                          validator.getValidUsernameForSignUp());
+                    }
+                }, usernameInput.getText().toString().trim());
+            }
+        });
+    }
```

Comment:
So, this is the implementation of the `onCallback` method of the new interface in this PR. 

This is needed because we need to **wait** until the firebase operation to check if the username is unique is **completed**. Firebase operations are **asynchronous**and occur in separate execution threads. This way, we can ensure that the username check has finished and we can proceed from there.